### PR TITLE
python-common/ceph/deployment: filter drives by actuators when creating osds

### PR DESCRIPTION
**Pull request by mjenglish**
As storage capacities grow, multi-actuator technology introduced by Seagate addresses the downward pressure on performance that comes with growing drive capacities and areal densities. Having multiple actuators enables drives to maintain the performance needs of customers with data-intensive applications.

Our testing indicates that, for certain workloads, using one OSD per actuator increases performance over using one OSD per dual-actuator drive. Bandwidth for the combined OSDs can increase over 40% and IOPS can nearly double compared to the alternative.

Extending the change introduced to ceph-volume that detects drives with two or more actuators, this feature enables the administrator to set a simple configuration that creates two OSDs on any detected dual-actuator drives using the advanced OSD service specification.

```
$ cat ~/osd.yaml
service_type: osd
service_id: example_osd_spec
placement:
  host_pattern: '*'
spec:
  data_devices:
    actuators: 2
  osds_per_device: 2
$ ./bin/ceph orch apply -i ~/osd.yaml
...
```

Copyright (c) 2022 Seagate Technology LLC and/or its Affiliates

Signed-off-by: Michael English <michael.english@seagate.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/ceph-volume/ceph_volume/util/disk.py
Comment (phlogistonjohn) [inline src/ceph-volume/ceph_volume/util/disk.py:875]: Python nitpick: here the default value is a string type.  However, below if the directory exists the actuators value becomes an int type. It's better to stick to one type, such as ints only. If a default of zero here is not appropriate I think `None` would be a better value. I think this would also be more type annotation friendly if they get added to this code sometime later.

Comment (mjenglish) [reply src/ceph-volume/ceph_volume/util/disk.py:875]: @phlogistonjohn Thanks for the feedback. This code can only prove presence of multiple actuators, not their absence, so it doesn't make sense to use zero as a default value here. An empty string was my choice, but let me think about your suggestion of using `None` instead. I'll have to test that.

You may also be interested in: https://github.com/ceph/ceph/pull/46344 .

**Comment by mjenglish** [comment]
@phlogistonjohn Would you like me to open a ticket in the orchestrator issue tracker for this pull request?

File: src/ceph-volume/ceph_volume/util/disk.py
Comment (phlogistonjohn) [reply src/ceph-volume/ceph_volume/util/disk.py:875]: In mypy speak, using None means the type of the value would be `Optional[int]` and is more typical rather than `Union[str, int]` where the only str value is ever "".

Ah, sorry I didn't notice that the patch was not unique to this PR. Next time if I have feedback on this patch I'll make on the original PR.

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn Would you like me to open a ticket in the orchestrator issue tracker for this pull request?

Personally, if the checklist doesn't require it I wouldn't. But this is more of an @adk3798 question. :-)

**Comment by adk3798** [comment]
> > @phlogistonjohn Would you like me to open a ticket in the orchestrator issue tracker for this pull request?
> 
> Personally, if the checklist doesn't require it I wouldn't. But this is more of an @adk3798 question. :-)

@mjenglish  I'm not going to require it, but it would be more convenient for us if a tracker was opened. Will also add I'd like @guits to review this before we merge, but he'll be away for a couple weeks. Hopefully this isn't super time sensitive.

**Comment by mjenglish** [comment]
@guits Would you like me to modify this code according to @phlogistonjohn 's feedback? If that is the case, I'd be happy to do so.

File: src/ceph-volume/ceph_volume/util/disk.py
Comment (mjenglish) [reply src/ceph-volume/ceph_volume/util/disk.py]: @phlogistonjohn I made the change. Do you have any other feedback for me to get this merged?

**Review by phlogistonjohn — APPROVED** [review]
LGTM, from a python programming pov.  I'll defer to the subject matter experts as for the functionality :-)

**Comment by guits** [comment]
jenkins test make check

**Review by adk3798 — CHANGES_REQUESTED** [review]

File: src/python-common/ceph/deployment/drive_group.py
Comment (mjenglish) [reply src/python-common/ceph/deployment/drive_group.py]: Thanks for the feedback. I fixed that specific issue, but make check is still failing and I don't see an obvious cause.

Comment (adk3798) [reply src/python-common/ceph/deployment/drive_group.py]: from the jenkins logs I'm seeing
```
ceph/deployment/drive_group.py:72:70: E502 the backslash is redundant between brackets
ceph/deployment/drive_group.py:73:17: E128 continuation line under-indented for visual indent
```
you should be able to go the ceph/src/python-common directory and run `tox -e lint` to try this locally as well I believe.

Comment (mjenglish) [reply src/python-common/ceph/deployment/drive_group.py]: It's fixed now.

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
will include this in next test run

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-10-05_13:02:39-orch:cephadm-wip-adk-testing-2022-10-05-0715-distro-default-smithi/

2 failures:

- 1 seemingly infra thing installing an selinux package https://tracker.ceph.com/issues/57771
- Failure in initial part of staggered upgrade while manually redeploying the mgr daemons with new images. This will need more investigation but I don't think any of the PRs in the run were necessarily causing the issue.

**Pull request merged**
